doc/design/switch_kernel.md

```diff
@@ -0,0 +1,66 @@
+
+The `layout` is useful for some computational library. One example is that MKLDNN uses many kinds of layout, such as `nChw8`. Each kind of layout will invoke the different kernel.
+
+## Problem
+
+We register a kernel for every operator and every kernel type ideally. However, it is impracticable for the following situations.
+
+1. Some operators, like CRF, are complicated and inefficient to be implemented on GPU. The CRF operator will only have a CPU kernel.
+2. Some operators will take too many memory. It is better to force them into CPU. However, the rest of operators in this neural network will be performed in GPU, i.e., model parallel problem.
+3. Some layout and place are particular. One example is that MKLDNN uses `nChw8` and there is no other library uses `nChw8`.
+
+Problems under these situations are similar. We can formalise this problem as follow.
+
+We register kernels with types $KT = \{kt_1, kt_2, kt_3, ...\}$ for one operator. The inputs of this operator should be run on kernel type $kt_{?}$, which the $kt_{?} \notin KT$. How to cast the input of this operator from $kt_{?}$ to any of kernel type in $KT$.
+
+## Solution
+
+It is clearly that transforming inputs of an operator from a specific kernel type to another kernel type is not related to the particular operator. So we should register these transformation methods as global methods.
+
+We can infer a kernel type from the inputs of an operators. We let this kernel type as `actual kernel type`, which means this kernel type is the actually kernel type that operator should be performed.
+
+We can get a kernel type by 1) The configuration of operator description. (Users may want to force use `MKL` for `conv` operator). 2) The place of the current executor. (Executor is running on GPU). This kernel type is what we expect the operator will be performed on. We let this kernel type as `expect kernel type`.
+
+We transform the input data from `actual` to `expect` if the expect kernel type is not as same as actual kernel type.
+
+The algorithm is described as follow
+
+```cpp
+using DataTransformationFN = std::function<void(const Tensor& in, Tensor* out)>;
+using KernelTypePair = std::pair<KernelType, KernelType>;
+
+map<KernelTypePair, DataTransformationFN> g_data_transformation_;
+
+void OpWithKernel::Run() {
+  vec<Tensor> inputs = ...
+  auto actual_kernel_type = GetActualKernelType(inputs);
+  
+  // The expected kernel type is related to actual kernel type.
+  // For the most operators, the expected kernel type is as same as
+  // actual kernel type.
+  //
+  // So we pass `actual_kernel_type` as a parameter of 
+  // GetExpectedKernelType
+  auto expect_kernel_type = GetExpectedKernelType(actual_kernel_type);
```

Comment:
Maybe `kernel.GetExpectedKernelType` ?

Reply:
Every operator instance has an expected kernel type. It is not an attribute in the kernel instance.

Reply:
The reason why I wrote like this is: (maybe it's just my concern)
1. the expected layout that are determined by current setting (eg. conv settings) and the platform (AVX2 or AVX512) **inside** the mkldnn kernel.

2. Is that possible that layout itself could be an attribute of input tensor not this kernel, so in the device who want to change the layout can just get from input tensor. No need to register may kernel types?

Reply:
1. The layout is an attribute of input tensor.
2. Switching kernel is not a job of the kernel itself. In Fluid, the kernel has only one method, `Kernel::Compute`. Switching kernel is not only used by `Conv` operator which uses `MKLDNN` as its backend but also used by `Pool(MKLDNN backend)` operator and many other operators. It is a higher level job. Because operators get and invoke `Kernel::Compute`. It is reasonable to let operators handle kernel switching.
3. `No need to register may kernel types?` All of these kernel types can be registered by a template class. For developers, they can just write one class and register it to every kernel type.
4. > the expected layout are determined by current setting (eg. conv settings) and the platform (AVX2 or AVX512) inside the mkldnn kernel.

    Yes. If Fluid is compiled with MKLDNN, the logic `GetExpectedKernelType` of `conv` and `pool` is just like you wrote.